# Importing, parsing, processing and exporting datasets (with the Pandas module)
**FIZ353 - Numerical Analysis | 09/10/2020**

    Importing a CSV file with Pandas
        Getting rid of the unnecessary header stuff
        Designating the column labels
        Designating the index using one of the columns
    Accessing to the data
        Old-school style: referring via the col/row indexes
        Referring via the col/row labels
    Filtering
    Processing
    Creating a DataFrame
    Exporting the DataFrame to a CSV file


Emre S. Tasci <emre.tasci@hacettepe.edu.tr>

Numerical analysis deals with various kinds of data: be it experimental results, surveys, collected values, etc. But one thing is almost always sure: there's always a huge amount of data lying around and we will be trying to make the most of it.

The first step to data processing is, of course to introduce the data to our code via importing (load).

Most of the time, the data is collected in matrix form and stored in comma-separated value (CSV) form, e.g.,

The data above is taken from [meteoblue](https://www.meteoblue.com/en/weather/historyclimate/weatherarchive/ankara_turkey_323786) site and contains the temperature, humidity and precipation data for Ankara for this month so far (8 days). But as you can observe, the actual data begins on the 11th line, with the previous lines containing identification information like location, coordinates, units, etc. This kind of preinformation lines are reffered to as _headers_.

## Importing a CSV file with Pandas

Importing the data in a CSV file is pretty straight-forward using Pandas' [`read_csv`](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.read_csv.html) command:

In [1]:
import pandas as pd
data_with_headerclutter = pd.read_csv("01_dataexport_20201008T180753.csv")
print(data_with_headerclutter)

          location       Ankara           Ankara.1             Ankara.2
0              lat     40.01502           40.01502             40.01502
1              lon     32.73532           32.73532             32.73532
2              asl      923.411            923.411              923.411
3         variable  Temperature  Relative Humidity  Precipitation Total
4             unit           °C                  %                   mm
..             ...          ...                ...                  ...
196  20201008T1900    26.437168               20.0                  0.0
197  20201008T2000     25.56717               21.0                  0.0
198  20201008T2100    24.777168               22.0                  0.0
199  20201008T2200    23.627169               24.0                  0.0
200  20201008T2300    22.727169               27.0                  0.0

[201 rows x 4 columns]


Here we see a couple of things, the foremost important one being that, Pandas don't worry about different types of variables as NumPy would. It just takes whatever it finds and -for the moment- treating the first line as the header row, takes all in.

Second, when we ask it to display the imported data, it prints a summary of things, 5 lines from the top and 5 lines from the bottom, also reporting the true size \[201 x 4\]. We can change the number of lines shown by setting the 'display.min_rows' & 'display.max_rows' options:

In [2]:
## Display a total of 6 rows only
#pd.set_option('display.min_rows', 6)
pd.set_option('display.max_rows', 6)
print(data_with_headerclutter)
## We can as well print it out by direct referral:
#data_with_headers

          location     Ankara  Ankara.1  Ankara.2
0              lat   40.01502  40.01502  40.01502
1              lon   32.73532  32.73532  32.73532
2              asl    923.411   923.411   923.411
..             ...        ...       ...       ...
198  20201008T2100  24.777168      22.0       0.0
199  20201008T2200  23.627169      24.0       0.0
200  20201008T2300  22.727169      27.0       0.0

[201 rows x 4 columns]


In [3]:
data_with_headerclutter.shape

(201, 4)

In [4]:
## Display all rows 
pd.set_option('display.max_rows', None)
print(data_with_headerclutter)

          location                                        Ankara  \
0              lat                                      40.01502   
1              lon                                      32.73532   
2              asl                                       923.411   
3         variable                                   Temperature   
4             unit                                            °C   
5            level                       2 m elevation corrected   
6       resolution                                        hourly   
7      aggregation                                          None   
8        timestamp  Ankara Temperature [2 m elevation corrected]   
9    20201001T0000                                     12.437169   
10   20201001T0100                                     12.557169   
11   20201001T0200                                     13.177169   
12   20201001T0300                                     13.087169   
13   20201001T0400                              

### Getting rid of the unnecessary header stuff
The headers are getting in the way (especially the 10th line ("timestamp | Ankara Temp...") even messes the output due to its enormous width. So we simply tell Pandas to start parsing from the 11th line (10th on the zero-indexed, thus we will be skipping 9 lines):

In [5]:
import pandas as pd
pd.set_option('display.min_rows', 10)
pd.set_option('display.max_rows', 10)
data_without_headerclutter = pd.read_csv("01_dataexport_20201008T180753.csv",
                                         skiprows=9)
data_without_headerclutter

timestamp  Ankara Temperature [2 m elevation corrected]  \
0    20201001T0000                                     12.437169   
1    20201001T0100                                     12.557169   
2    20201001T0200                                     13.177169   
3    20201001T0300                                     13.087169   
4    20201001T0400                                     12.867169   
..             ...                                           ...   
187  20201008T1900                                     26.437168   
188  20201008T2000                                     25.567170   
189  20201008T2100                                     24.777168   
190  20201008T2200                                     23.627169   
191  20201008T2300                                     22.727169   

     Ankara Relative Humidity [2 m]  Ankara Precipitation Total  
0                              59.0                         0.0  
1                              63.0                         0.0  
2                              68.0                         0.0  
3                              76.0                         0.0  
4                              81.0                         0.0  
..                              ...                         ...  
187                            20.0                         0.0  
188                            21.0                         0.0  
189                            22.0                         0.0  
190                            24.0                         0.0  
191                            27.0                         0.0  

[192 rows x 4 columns]

### Designating the column labels
... almost there but it has assumed the previous line before that as the column label line (which is usually true but this time there is just too much clutter), so we'll manually enter the column labels:

In [6]:
data_without_headerclutter.columns = ['Timestamp','Temperature','Relative Humidity',
                                      'Precipitation Total']
print(data_without_headerclutter)

         Timestamp  Temperature  Relative Humidity  Precipitation Total
0    20201001T0000    12.437169               59.0                  0.0
1    20201001T0100    12.557169               63.0                  0.0
2    20201001T0200    13.177169               68.0                  0.0
3    20201001T0300    13.087169               76.0                  0.0
4    20201001T0400    12.867169               81.0                  0.0
..             ...          ...                ...                  ...
187  20201008T1900    26.437168               20.0                  0.0
188  20201008T2000    25.567170               21.0                  0.0
189  20201008T2100    24.777168               22.0                  0.0
190  20201008T2200    23.627169               24.0                  0.0
191  20201008T2300    22.727169               27.0                  0.0

[192 rows x 4 columns]


In [7]:
data_without_headerclutter

Timestamp  Temperature  Relative Humidity  Precipitation Total
0    20201001T0000    12.437169               59.0                  0.0
1    20201001T0100    12.557169               63.0                  0.0
2    20201001T0200    13.177169               68.0                  0.0
3    20201001T0300    13.087169               76.0                  0.0
4    20201001T0400    12.867169               81.0                  0.0
..             ...          ...                ...                  ...
187  20201008T1900    26.437168               20.0                  0.0
188  20201008T2000    25.567170               21.0                  0.0
189  20201008T2100    24.777168               22.0                  0.0
190  20201008T2200    23.627169               24.0                  0.0
191  20201008T2300    22.727169               27.0                  0.0

[192 rows x 4 columns]

In [8]:
data_without_headerclutter.iloc[[2],[0,1,2,3]]

Timestamp  Temperature  Relative Humidity  Precipitation Total
2  20201001T0200    13.177169               68.0                  0.0

### Designating the index using one of the columns
Currently, the index (row ids) is enumerated. But if we wish, we can also designate a column that holds unique values (such as the `Timestamp` column in our example) via the `set_index` command:

In [9]:
df2 = data_without_headerclutter.set_index('Timestamp')
df2

Temperature  Relative Humidity  Precipitation Total
Timestamp                                                         
20201001T0000    12.437169               59.0                  0.0
20201001T0100    12.557169               63.0                  0.0
20201001T0200    13.177169               68.0                  0.0
20201001T0300    13.087169               76.0                  0.0
20201001T0400    12.867169               81.0                  0.0
...                    ...                ...                  ...
20201008T1900    26.437168               20.0                  0.0
20201008T2000    25.567170               21.0                  0.0
20201008T2100    24.777168               22.0                  0.0
20201008T2200    23.627169               24.0                  0.0
20201008T2300    22.727169               27.0                  0.0

[192 rows x 3 columns]

In [10]:
data_without_headerclutter

Timestamp  Temperature  Relative Humidity  Precipitation Total
0    20201001T0000    12.437169               59.0                  0.0
1    20201001T0100    12.557169               63.0                  0.0
2    20201001T0200    13.177169               68.0                  0.0
3    20201001T0300    13.087169               76.0                  0.0
4    20201001T0400    12.867169               81.0                  0.0
..             ...          ...                ...                  ...
187  20201008T1900    26.437168               20.0                  0.0
188  20201008T2000    25.567170               21.0                  0.0
189  20201008T2100    24.777168               22.0                  0.0
190  20201008T2200    23.627169               24.0                  0.0
191  20201008T2300    22.727169               27.0                  0.0

[192 rows x 4 columns]

In [11]:
data_without_headerclutter = data_without_headerclutter.set_index('Timestamp')
data_without_headerclutter

Temperature  Relative Humidity  Precipitation Total
Timestamp                                                         
20201001T0000    12.437169               59.0                  0.0
20201001T0100    12.557169               63.0                  0.0
20201001T0200    13.177169               68.0                  0.0
20201001T0300    13.087169               76.0                  0.0
20201001T0400    12.867169               81.0                  0.0
...                    ...                ...                  ...
20201008T1900    26.437168               20.0                  0.0
20201008T2000    25.567170               21.0                  0.0
20201008T2100    24.777168               22.0                  0.0
20201008T2200    23.627169               24.0                  0.0
20201008T2300    22.727169               27.0                  0.0

[192 rows x 3 columns]

and thus we now have 3 columns with the former `Timestamp` column being the index (row identifier).

We can get the list of the column names and row indexes:

In [12]:
# Columns
data_without_headerclutter.columns

Index(['Temperature', 'Relative Humidity', 'Precipitation Total'], dtype='object')

In [13]:
# Row indexes
data_without_headerclutter.index

Index(['20201001T0000', '20201001T0100', '20201001T0200', '20201001T0300',
       '20201001T0400', '20201001T0500', '20201001T0600', '20201001T0700',
       '20201001T0800', '20201001T0900',
       ...
       '20201008T1400', '20201008T1500', '20201008T1600', '20201008T1700',
       '20201008T1800', '20201008T1900', '20201008T2000', '20201008T2100',
       '20201008T2200', '20201008T2300'],
      dtype='object', name='Timestamp', length=192)

## Accessing to the data
Now that we have managed to import the datafile, we have all the data under our reach - hooray! 8)

From here on, it runs more or less in the same vein as a NumPy or GNU Octave/MATLAB arrays, with the main difference being the ability to also refer directly using the column and row labels.

### Old-school style: referring via the col/row indexes
For this kind of referrence, we use the `iloc` command. But before we do that, let me redefine the dataset name to `df` (for 'DataFrame').

In [14]:
df = data_without_headerclutter 
# (I'm just tired of typing 'data_without_headerclutter' all the time!)

In [15]:
df

Temperature  Relative Humidity  Precipitation Total
Timestamp                                                         
20201001T0000    12.437169               59.0                  0.0
20201001T0100    12.557169               63.0                  0.0
20201001T0200    13.177169               68.0                  0.0
20201001T0300    13.087169               76.0                  0.0
20201001T0400    12.867169               81.0                  0.0
...                    ...                ...                  ...
20201008T1900    26.437168               20.0                  0.0
20201008T2000    25.567170               21.0                  0.0
20201008T2100    24.777168               22.0                  0.0
20201008T2200    23.627169               24.0                  0.0
20201008T2300    22.727169               27.0                  0.0

[192 rows x 3 columns]

In [16]:
df.iloc[2,0] # Row 2, Col 0

13.177169000000001

In [17]:
df.iloc[[4,1],[1,0]] # Rows 4 and 1 && Cols 1 and 0

Relative Humidity  Temperature
Timestamp                                    
20201001T0400               81.0    12.867169
20201001T0100               63.0    12.557169

In [18]:
df.iloc[1:4] # Rows [1,4) && Cols [0,2)

Temperature  Relative Humidity  Precipitation Total
Timestamp                                                         
20201001T0100    12.557169               63.0                  0.0
20201001T0200    13.177169               68.0                  0.0
20201001T0300    13.087169               76.0                  0.0

In [19]:
df.iloc[[1,3,6],:] # Rows 1,3 and 6 && All cols

Temperature  Relative Humidity  Precipitation Total
Timestamp                                                         
20201001T0100    12.557169               63.0                  0.0
20201001T0300    13.087169               76.0                  0.0
20201001T0600    11.177169               90.0                  0.0

### Referring via the col/row labels
And then, we have the option to call by the labels, using `loc`:

In [20]:
df.loc[['20201001T0100'],['Temperature']]

Temperature
Timestamp                 
20201001T0100    12.557169

In [21]:
df.loc[['20201001T0300'],['Temperature','Precipitation Total']]

Temperature  Precipitation Total
Timestamp                                      
20201001T0300    13.087169                  0.0

# Filtering
For filtering, we just make a proposition, and get the True/False Boolean results. Let's work with a smaller dataframe:

In [22]:
sdf = df.iloc[0:5,0:2]
sdf

Temperature  Relative Humidity
Timestamp                                    
20201001T0000    12.437169               59.0
20201001T0100    12.557169               63.0
20201001T0200    13.177169               68.0
20201001T0300    13.087169               76.0
20201001T0400    12.867169               81.0

Let's try to find those entries with temperature below 13 degrees:

In [23]:
sdf.iloc[[2,4]]

Temperature  Relative Humidity
Timestamp                                    
20201001T0200    13.177169               68.0
20201001T0400    12.867169               81.0

In [24]:
sdf.iloc[:,[0]]

Temperature
Timestamp                 
20201001T0000    12.437169
20201001T0100    12.557169
20201001T0200    13.177169
20201001T0300    13.087169
20201001T0400    12.867169

In [25]:
sdf.iloc[2]

Temperature          13.177169
Relative Humidity    68.000000
Name: 20201001T0200, dtype: float64

In [26]:
sdf.iloc[:,:].values

array([[12.437169, 59.      ],
       [12.557169, 63.      ],
       [13.177169, 68.      ],
       [13.087169, 76.      ],
       [12.867169, 81.      ]])

In [27]:
sdf.values

array([[12.437169, 59.      ],
       [12.557169, 63.      ],
       [13.177169, 68.      ],
       [13.087169, 76.      ],
       [12.867169, 81.      ]])

In [28]:
sdf.iloc[:,[0]] < 13

Temperature
Timestamp                 
20201001T0000         True
20201001T0100         True
20201001T0200        False
20201001T0300        False
20201001T0400         True

So we see that for 3 of the 5 entries, this assertion is correct, let's pick them:

In [29]:
filter1 = sdf.iloc[:,[0]] < 13
sdf.loc[filter1['Temperature']==True]

Temperature  Relative Humidity
Timestamp                                    
20201001T0000    12.437169               59.0
20201001T0100    12.557169               63.0
20201001T0400    12.867169               81.0

In [30]:
filter1

Temperature
Timestamp                 
20201001T0000         True
20201001T0100         True
20201001T0200        False
20201001T0300        False
20201001T0400         True

In [31]:
filter1.values

array([[ True],
       [ True],
       [False],
       [False],
       [ True]])

In [32]:
sdf[filter1.values]

Temperature  Relative Humidity
Timestamp                                    
20201001T0000    12.437169               59.0
20201001T0100    12.557169               63.0
20201001T0400    12.867169               81.0

We can achieve the same thing by going straight to the heart of the issue using the column name directly:

In [33]:
sdf.Temperature

Timestamp
20201001T0000    12.437169
20201001T0100    12.557169
20201001T0200    13.177169
20201001T0300    13.087169
20201001T0400    12.867169
Name: Temperature, dtype: float64

In [34]:
sdf.Temperature < 13

Timestamp
20201001T0000     True
20201001T0100     True
20201001T0200    False
20201001T0300    False
20201001T0400     True
Name: Temperature, dtype: bool

and then feeding this as the index:

In [35]:
sdf[sdf.Temperature < 13]

Temperature  Relative Humidity
Timestamp                                    
20201001T0000    12.437169               59.0
20201001T0100    12.557169               63.0
20201001T0400    12.867169               81.0

What about the row index? Can we specify a criteria for them as well? First let's re-display our small dataframe:

In [36]:
sdf

Temperature  Relative Humidity
Timestamp                                    
20201001T0000    12.437169               59.0
20201001T0100    12.557169               63.0
20201001T0200    13.177169               68.0
20201001T0300    13.087169               76.0
20201001T0400    12.867169               81.0

In [37]:
sdf.columns

Index(['Temperature', 'Relative Humidity'], dtype='object')

In [38]:
sdf.index

Index(['20201001T0000', '20201001T0100', '20201001T0200', '20201001T0300',
       '20201001T0400'],
      dtype='object', name='Timestamp')

Let's seek the ones recorded before (and including) 01:00 **or** the one at 04:00 (here _or_ is for the union, not intersection!)

In [39]:
# Joining multiple criteria based on row label (index)
sdf[(sdf.index < '20201001T0200') | (sdf.index == '20201001T0400')]

Temperature  Relative Humidity
Timestamp                                    
20201001T0000    12.437169               59.0
20201001T0100    12.557169               63.0
20201001T0400    12.867169               81.0

As the row index is string, we don't even need to write the whole timestamp to the end to make the comparison:

In [40]:
sdf.index > "20201001T02"

array([False, False,  True,  True,  True])

In [41]:
sdf[sdf.index> "20201001T02"]

Temperature  Relative Humidity
Timestamp                                    
20201001T0200    13.177169               68.0
20201001T0300    13.087169               76.0
20201001T0400    12.867169               81.0

In [42]:
sdf.loc[['20201001T0300'],:]

Temperature  Relative Humidity
Timestamp                                    
20201001T0300    13.087169               76.0

## Processing
Now that we know how to slice via filtering, we can do whatever we want with the sections of the data we're interested in. For example, let us calculate the mean temperature in October, 1st:

In [43]:
df

Temperature  Relative Humidity  Precipitation Total
Timestamp                                                         
20201001T0000    12.437169               59.0                  0.0
20201001T0100    12.557169               63.0                  0.0
20201001T0200    13.177169               68.0                  0.0
20201001T0300    13.087169               76.0                  0.0
20201001T0400    12.867169               81.0                  0.0
...                    ...                ...                  ...
20201008T1900    26.437168               20.0                  0.0
20201008T2000    25.567170               21.0                  0.0
20201008T2100    24.777168               22.0                  0.0
20201008T2200    23.627169               24.0                  0.0
20201008T2300    22.727169               27.0                  0.0

[192 rows x 3 columns]

In [44]:
# First get the October 1st day's Temperature data:
oct1_temp = df[df.index < "20201002"]['Temperature']
print(oct1_temp)

Timestamp
20201001T0000    12.437169
20201001T0100    12.557169
20201001T0200    13.177169
20201001T0300    13.087169
20201001T0400    12.867169
                   ...    
20201001T1900    18.537169
20201001T2000    16.797169
20201001T2100    14.717169
20201001T2200    12.627169
20201001T2300    11.187169
Name: Temperature, Length: 24, dtype: float64


In [45]:
oct1_temp.mean()

16.406335583333334

Let's calculate all the daily mean temperatures:

In [46]:
"Hello"+str(3)

'Hello3'

In [47]:
for i in range(1,9):
    day_temp = df[(df.index > "2020100"+str(i-1)) & (df.index < "2020100"+str(i+1))]['Temperature']
    day_temp_mean = day_temp.mean()
    print ("October, {} mean temperature: {:6.3f} degree C".format(i,day_temp_mean))

October, 1 mean temperature: 16.406 degree C
October, 2 mean temperature: 15.769 degree C
October, 3 mean temperature: 16.272 degree C
October, 4 mean temperature: 17.630 degree C
October, 5 mean temperature: 18.854 degree C
October, 6 mean temperature: 21.028 degree C
October, 7 mean temperature: 22.930 degree C
October, 8 mean temperature: 23.917 degree C


# Creating a dataframe
We can directly create a dataframe with the `DataFrame` command:

In [48]:
# Start with an empty dataframe:
mean_temps = pd.DataFrame({'day' : [],'meanTemp' : []})
mean_temps

Empty DataFrame
Columns: [day, meanTemp]
Index: []

In [49]:
# We can fill it individually
mean_temps.append({'day': "20201001", 'meanTemp' : 16.41}, ignore_index=True)

day  meanTemp
0  20201001     16.41

In [50]:
# or create it from the existing list
days = []
means = []
for i in range(1,9):
    day_temp = df[(df.index > "2020100"+str(i-1)) & (df.index < "2020100"+str(i+1))]\
        ['Temperature']
    day_temp_mean = day_temp.mean()
    # print ("October, {} mean temperature: {:5.2f} degree C".format(i,day_temp_mean))
    days.append("2020100"+str(i))
    means.append(day_temp_mean)
# ------------------------------
print(days)
print(means)
data = {'days': days, 'means': means}
df_mean_temps = pd.DataFrame(data)
df_mean_temps

['20201001', '20201002', '20201003', '20201004', '20201005', '20201006', '20201007', '20201008']
[16.406335583333334, 15.769460629166666, 16.271544045833334, 17.629669104166666, 18.854460770833334, 21.028210770833333, 22.93029415625, 23.916752552083334]


days      means
0  20201001  16.406336
1  20201002  15.769461
2  20201003  16.271544
3  20201004  17.629669
4  20201005  18.854461
5  20201006  21.028211
6  20201007  22.930294
7  20201008  23.916753

In [51]:
new_entry = pd.Series({'days': "20201013",'means':"23.5"})
df_mean_temps = df_mean_temps.append(new_entry, ignore_index=True)
df_mean_temps

days    means
0  20201001  16.4063
1  20201002  15.7695
2  20201003  16.2715
3  20201004  17.6297
4  20201005  18.8545
5  20201006  21.0282
6  20201007  22.9303
7  20201008  23.9168
8  20201013     23.5

In [52]:
# Specifying the index:
new_entry = pd.Series({'days': "20201011",'means':"24.5"}, name = 'k14')
df_mean_temps = df_mean_temps.append(new_entry)

In [53]:
df_mean_temps

days    means
0    20201001  16.4063
1    20201002  15.7695
2    20201003  16.2715
3    20201004  17.6297
4    20201005  18.8545
5    20201006  21.0282
6    20201007  22.9303
7    20201008  23.9168
8    20201013     23.5
k14  20201011     24.5

In [54]:
df_mean_temps.index

Index([0, 1, 2, 3, 4, 5, 6, 7, 8, 'k14'], dtype='object')

In [55]:
df_mean_temps.shape

(10, 2)

In [56]:
df_mean_temps.iloc[[9],:]

days means
k14  20201011  24.5

In [57]:
df_mean_temps.loc[['k14'],:]

days means
k14  20201011  24.5

In [58]:
df_mean_temps.iloc[[8],:]

days means
8  20201013  23.5

In [59]:
df_mean_temps.loc[[8],:]

days means
8  20201013  23.5

# Exporting the DataFrame to a CSV file


In [60]:
df_mean_temps

days    means
0    20201001  16.4063
1    20201002  15.7695
2    20201003  16.2715
3    20201004  17.6297
4    20201005  18.8545
5    20201006  21.0282
6    20201007  22.9303
7    20201008  23.9168
8    20201013     23.5
k14  20201011     24.5

In [61]:
# With the row labels (index) and column labels included:
df_mean_temps.to_csv('data/01_out.csv')
print(df_mean_temps.to_csv())

,days,means
0,20201001,16.406335583333334
1,20201002,15.769460629166666
2,20201003,16.271544045833334
3,20201004,17.629669104166666
4,20201005,18.854460770833334
5,20201006,21.028210770833333
6,20201007,22.93029415625
7,20201008,23.916752552083334
8,20201013,23.5
k14,20201011,24.5



In [62]:
# Without the row labels but with the column labels:
print(df_mean_temps.to_csv(index=False))

days,means
20201001,16.406335583333334
20201002,15.769460629166666
20201003,16.271544045833334
20201004,17.629669104166666
20201005,18.854460770833334
20201006,21.028210770833333
20201007,22.93029415625
20201008,23.916752552083334
20201013,23.5
20201011,24.5



In [63]:
# Without the row labels and the column labels:
print(df_mean_temps.to_csv(index=False, header=False))

20201001,16.406335583333334
20201002,15.769460629166666
20201003,16.271544045833334
20201004,17.629669104166666
20201005,18.854460770833334
20201006,21.028210770833333
20201007,22.93029415625
20201008,23.916752552083334
20201013,23.5
20201011,24.5



In [64]:
# Specify header for the index column:
print(df_mean_temps.to_csv(index_label='ids'))

ids,days,means
0,20201001,16.406335583333334
1,20201002,15.769460629166666
2,20201003,16.271544045833334
3,20201004,17.629669104166666
4,20201005,18.854460770833334
5,20201006,21.028210770833333
6,20201007,22.93029415625
7,20201008,23.916752552083334
8,20201013,23.5
k14,20201011,24.5



In [65]:
# Use ';' as the seperator, instead of ','
df_mean_temps.to_csv(sep=';',path_or_buf='data/01_out.csv')
print(df_mean_temps.to_csv(sep=';'))

;days;means
0;20201001;16.406335583333334
1;20201002;15.769460629166666
2;20201003;16.271544045833334
3;20201004;17.629669104166666
4;20201005;18.854460770833334
5;20201006;21.028210770833333
6;20201007;22.93029415625
7;20201008;23.916752552083334
8;20201013;23.5
k14;20201011;24.5



# NumPy way of doing things
Even though Pandas offer very flexible options of doing things, we could as well have used NumPy:

In [66]:
import numpy as np

data = np.genfromtxt("data/01_dataexport_20201008T180753.csv", delimiter=',')
print(data)

[[        nan         nan         nan         nan]
 [        nan  40.01502    40.01502    40.01502  ]
 [        nan  32.73532    32.73532    32.73532  ]
 [        nan 923.411     923.411     923.411    ]
 [        nan         nan         nan         nan]
 [        nan         nan         nan         nan]
 [        nan         nan         nan         nan]
 [        nan         nan         nan         nan]
 [        nan         nan         nan         nan]
 [        nan         nan         nan         nan]
 [        nan  12.437169   59.          0.       ]
 [        nan  12.557169   63.          0.       ]
 [        nan  13.177169   68.          0.       ]
 [        nan  13.087169   76.          0.       ]
 [        nan  12.867169   81.          0.       ]
 [        nan  11.567169   88.          0.       ]
 [        nan  11.177169   90.          0.       ]
 [        nan  12.187169   87.          0.       ]
 [        nan  13.797169   78.          0.       ]
 [        nan  14.967169   72. 

As you can see, unfortunately, we can't place different types of variables into the same array. But other than that, the rest of the operations are similar:

In [67]:
data.shape

(202, 4)

In [68]:
data_crop = data[11:,1:4]

In [69]:
data_crop

array([[12.557169 , 63.       ,  0.       ],
       [13.177169 , 68.       ,  0.       ],
       [13.087169 , 76.       ,  0.       ],
       [12.867169 , 81.       ,  0.       ],
       [11.567169 , 88.       ,  0.       ],
       [11.177169 , 90.       ,  0.       ],
       [12.187169 , 87.       ,  0.       ],
       [13.797169 , 78.       ,  0.       ],
       [14.967169 , 72.       ,  0.       ],
       [16.787169 , 62.       ,  0.       ],
       [18.367168 , 55.       ,  0.       ],
       [20.957169 , 41.       ,  0.       ],
       [22.057169 , 36.       ,  0.       ],
       [22.857168 , 33.       ,  0.       ],
       [23.527168 , 29.       ,  0.       ],
       [23.40717  , 28.       ,  0.       ],
       [22.687168 , 28.       ,  0.       ],
       [21.41717  , 30.       ,  0.       ],
       [18.537169 , 35.       ,  0.       ],
       [16.797169 , 39.       ,  0.       ],
       [14.717169 , 44.       ,  0.       ],
       [12.627169 , 54.       ,  0.       ],
       [11

In [70]:
np.mean(data_crop[:,0])

19.623451820418843